### DISCUSSION TITLE
Git push and envoyer

### DISCUSSION THREAD

**Discussion by brew-derek** [statamic/cms#6844] [Troubleshooting]
Setting up the git automation for the first time with a new licence and followed these [instructions](https://statamic.dev/tips/zero-downtime-deployments#git-automation) to setup the hooks on Envoyer correctly, however when it comes to this:
```
git branch --track master origin/master
```
and I've changed master to `main` the hook then doesn't run and gives me back an error like this:
```
 * [new branch]      main       -> origin/main
Branch 'main' set up to track remote branch 'main' from 'origin'.
fatal: ambiguous argument 'HEAD': unknown revision or path not in the working tree.
Use '--' to separate paths from revisions, like this:
'git <command> [<revision>...] -- [<file>...]'
```

I tried keeping the branch on the server called `master` and tracking that to `origin/main` which doesn't give me an error but then nothing is automatically pushed (`STATAMIC_GIT_PUSH=true` is set in the .env) although the commits are all made as expected. If I try manually pushing I get this
```
git push
fatal: The upstream branch of your current branch does not match
the name of your current branch.  To push to the upstream branch
on the remote, use

    git push origin HEAD:main

To push to the branch of the same name on the remote, use

    git push origin HEAD

To choose either option permanently, see push.default in 'git help config'.
```
If I run the first one as suggested it pushes perfectly fine.

I can only assume I am doing something wrong due to inexperience with the git automation? 

I have seen others having issues with push, in particular the protocol on the remote changing when deploying, but as far as I can tell mine is always `ssh`.

Anyone have any clue what's happening?

**Comment by edalzell** [comment]
Have you tried running each of those commands on the command line? I suspect there's a branch issue with the names etc.

**Reply by brew-derek** [reply]
Yeah, everything works fine apart from the push, as I'd changed the local tracking branch name it wants to run `git push origin HEAD:main` and not just a plain push.
If I SSH in and manually set the tracking branch then check it out, everything works as expected 
```
git branch --track main origin/main
git checkout main
```

Given that I think just adding the `git checkout main` onto the end of the git object creation hook in Envoyer seems to fix it, so that now looks like this for me:
```
cd {{ release }}
 
git init
git remote add origin git@github.com:myrepo/myrepo.git
git fetch
git branch --track main origin/main
git reset origin/main
git checkout main
```
Unless anyone thinks that looks stupid (it isn't in production yet) I am going to go with that?

**Reply by jesseleite** [reply]
@brew-derek We have a docs article on using Git with Envoyer here, which suggests pretty similar to what you came up with... https://statamic.dev/tips/zero-downtime-deployments#setting-up-your-git-remote-in-envoyer-releases

Definitely give the whole article a read though, as there are a few other gotchas when using Statamic's Git Automation with Envoyer as well.
